addons/nuxeo-retention/nuxeo-retention-core/src/main/java/org/nuxeo/retention/service/RetentionManagerImpl.java

```diff
@@ -0,0 +1,353 @@
+
+    @Override
+    public DocumentModel attachRule(DocumentModel document, RetentionRule rule, CoreSession session) {
+        checkCanAttachRule(document, rule, session);
+        session.makeRecord(document.getRef());
+        final Calendar retainUntil;
+        if (rule.isImmediate()) {
+            retainUntil = rule.getRetainUntilDateFromNow();
+            log.debug("Attaching immediate rule until {}",
+                    () -> RetentionConstants.DEFAULT_DATE_FORMAT.format(retainUntil.getTime()));
+        } else if (rule.isAfterDely()) {
+            log.debug("Attaching after delay rule");
+            throw new UnsupportedOperationException("After delay not yet implemented");
+        } else if (rule.isEventBased()) {
+            retainUntil = CoreSession.RETAIN_UNTIL_INDETERMINATE;
+            log.debug("Attaching event-based rule on {} matching \"{}\"", rule.getStartingPointEvent(),
+                    rule.getStartingPointExpression());
+        } else if (rule.isMetadataBased()) {
+            String xpath = rule.getMetadataXpath();
+            if (StringUtils.isBlank(xpath)) {
+                throw new NuxeoException("Metadata field is null");
+            }
+            Property prop = document.getProperty(xpath);
+            if (!(prop.getType() instanceof DateType)) {
+                throw new NuxeoException(
+                        String.format("Field %s of type % is expected to have a DateType", xpath, prop.getType()));
+            }
+            Calendar value = (Calendar) prop.getValue();
+            if (value == null) {
+                retainUntil = rule.getRetainUntilDateFromNow();
+            } else {
+                retainUntil = rule.getRetainUntilDateFrom(value);
+                Calendar now = Calendar.getInstance();
+                if (now.after(retainUntil)) {
+                    log.info(
+                            "Metabased-based rule found past date {} as retention expiration date on {} from {} property. Ignoring...",
+                            () -> RetentionConstants.DEFAULT_DATE_FORMAT.format(retainUntil.getTime()),
+                            document::getPathAsString, () -> xpath);
+                    return session.getDocument(document.getRef());
+                }
+            }
+            log.debug("Attaching rule base on {} with value {}", () -> xpath,
+                    () -> RetentionConstants.DEFAULT_DATE_FORMAT.format(retainUntil.getTime()));
+        } else {
+            throw new IllegalArgumentException("Unknown starting point policy: " + rule.getStartingPointPolicy());
+        }
+        document.addFacet(RetentionConstants.RECORD_FACET);
+        Record record = document.getAdapter(Record.class);
+        record.setRule(rule, session);
+        executeRuleBeginActions(record, session);
+        session.setRetainUntil(document.getRef(), retainUntil, null);
+        notifyAttachRule(record, rule, session);
+        return session.getDocument(document.getRef());
```

Comment:
i am wondering if we can rework this part to reduce the number of the control flow. perhaps separate the control/validation from the processing